app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt

```diff
@@ -0,0 +1,100 @@
+
+import android.content.Context
+import android.os.Bundle
+import android.print.PrintManager
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.webkit.WebView
+import androidx.core.view.get
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.viewModels
+import androidx.navigation.fragment.findNavController
+import androidx.navigation.fragment.navArgs
+import com.infomaniak.mail.R
+import com.infomaniak.mail.data.LocalSettings
+import com.infomaniak.mail.databinding.FragmentPrintMailBinding
+import dagger.hilt.android.AndroidEntryPoint
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class PrintMailFragment : Fragment() {
+
+    @Inject
+    lateinit var localSettings: LocalSettings
+
+    private var _binding: FragmentPrintMailBinding? = null
+    private val binding get() = _binding!! // This property is only valid between onCreateView and onDestroyView
+
+    private val navigationArgs: PrintMailFragmentArgs by navArgs()
+    private val threadViewModel: ThreadViewModel by viewModels()
+    private val threadAdapter inline get() = binding.messagesList.adapter as ThreadAdapter
+
+    private var subject = ""
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
+        return FragmentPrintMailBinding.inflate(inflater, container, false).also { _binding = it }.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?): Unit = with(threadViewModel) {
+        super.onViewCreated(view, savedInstanceState)
+        setupAdapter()
+        threadLive.observe(viewLifecycleOwner) { thread ->
+            thread?.subject?.let { subject = it }
+        }
+        messagesLive.observe(viewLifecycleOwner) { messages ->
+            threadAdapter.submitList(messages.filter { it.uid == navigationArgs.messageUid })
+        }
+        navigationArgs.openThreadUid?.let {
+            reassignThreadLive(it)
+            reassignMessagesLive(it)
+        }
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        _binding = null
+    }
+
+    override fun onPause() {
+        super.onPause()
+        findNavController().popBackStack()
+    }
+
+    private fun setupAdapter() = with(binding.messagesList) {
+        adapter = ThreadAdapter(
+            shouldLoadDistantResources = true,
+            isForPrinting = true,
+            onBodyWebviewFinishedLoading = {
+                CoroutineScope(Dispatchers.Main).launch {
+                    val webViewPrintAdapter = getWebViewToPrint().createPrintDocumentAdapter(subject)
+                    val printManager = requireContext().getSystemService(Context.PRINT_SERVICE) as PrintManager
+                    printManager.print(subject, webViewPrintAdapter, null)
+                }
```

Comment:
Why do we need to lunch a coroutine if it's to put it on the main thread?

Reply:
I have this error when doing so:

```
Java.lang.RuntimeException: java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. (Expected Looper Looper (main, tid 2) {ed6baa7} called on Looper (JavaBridge, tid 141) {bf80207}, FYI main Looper is Looper (main, tid 2) {ed6baa7})
2024-01-30 14:48:03.788  3292-5327  System.err              com.infomaniak.mail                  W  	at android.webkit.WebView.checkThread(WebView.java:2626)
2024-01-30 14:48:03.788  3292-5327  System.err              com.infomaniak.mail                  W  	at android.webkit.WebView.createPrintDocumentAdapter(WebView.java:1151)
2024-01-30 14:48:03.788  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.PrintMailFragment$setupAdapter$1$1.invoke(PrintMailFragment.kt:91)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.PrintMailFragment$setupAdapter$1$1.invoke(PrintMailFragment.kt:86)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.ThreadAdapter$bindBody$1$5.invoke(ThreadAdapter.kt:431)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.ThreadAdapter$bindBody$1$5.invoke(ThreadAdapter.kt:427)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.utils.WebViewUtils$JavascriptBridge.webviewFinishedLoading(WebViewUtils.kt:86)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.MessageQueue.nativePollOnce(Native Method)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.MessageQueue.next(MessageQueue.java:335)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.Looper.loopOnce(Looper.java:162)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.Looper.loop(Looper.java:294)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.HandlerThread.run(HandlerThread.java:67)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  Caused by: java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. (Expected Looper Looper (main, tid 2) {ed6baa7} called on Looper (JavaBridge, tid 141) {bf80207}, FYI main Looper is Looper (main, tid 2) {ed6baa7})
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.webkit.WebView.checkThread(WebView.java:2621)
```